strawberryfields/engine.py

```diff
@@ -277,15 +271,35 @@ def _broadcast_nones(val, dim):
                 p = p.compile(target, **compile_options)
             p.lock()
 
-            self._run_program(p, **kwargs)
-            shots = kwargs.get("shots", 1)
-            self.samples = [_broadcast_nones(p.reg_refs[k].val, shots) for k in sorted(p.reg_refs)]
+            _, values = self._run_program(p, **kwargs)
             self.run_progs.append(p)
 
+            if len(values) > 1:
+                sort_order = [
+                    r.ind for c in p.circuit for r in c.reg if "Measure" in c.op.__str__()
+                ]
+
+                # check for duplicate mode-measures
+                if len(sort_order) != len(set(sort_order)):
+                    raise RuntimeError("Modes can only be measured once inside a circuit.")
+
+                self.samples = Result.combine_samples(values, sort_order)
```

Comment:
could it be worth having `combine_samples` be a method of `Engine`?

Reply:
I'm not sure. I feel like the function has more to do with the results than the engine, so I think it might fit well inside `results.py`. @josh146, what do you think?

Reply:
Yep, I tend to agree, I feel that it fits in more as a static method of the results class (but it is a very fuzzy line, and could arguably go in both places).

On a tangent, I think it was an oversight that the Results class got moved into the API module, since it is more general than just the API client? 🤔